zaza/utilities/openstack_provider.py

```diff
@@ -0,0 +1,305 @@
+# Copyright 2021 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Functions to work with the openstack provider."""
+
+import logging
+import os
+import tenacity
+
+from keystoneauth1 import session
+from keystoneauth1.identity import (
+    v3,
+    v2,
+)
+from novaclient import client as novaclient_client
+
+
+class MissingOSAthenticationException(Exception):
+    """Exception when some data needed to authenticate is missing."""
+
+    pass
+
+
+def get_undercloud_keystone_session(verify=None):
+    """Return Under cloud keystone session.
+
+    :param verify: Control TLS certificate verification behaviour
+    :type verify: any
+    :returns keystone_session: keystoneauth1.session.Session object
+    :rtype: keystoneauth1.session.Session
+    """
+    return get_keystone_session(get_undercloud_auth(),
+                                verify=verify)
+
+
+def get_keystone_session(openrc_creds, scope='PROJECT', verify=None):
+    """Return keystone session.
+
+    :param openrc_creds: OpenStack RC credentials
```

Comment:
I find it strange to talk about `RC` as I think it's refering to an `*rc` file that you would source, but in fact what we're really talking about here are creds coming from env vars, not necessary from an `*rc` file, right? 

Feel free to leave as is if it's code that you copied from somewhere else and haven't authored yourself.

Reply:
Well the "R" is resource, and the "C" is credentials, so, it is a bit of a tautology!  But its not really referring to the file; more the concept of key/value pairs that make up the openrc 'set'.

I think I'd like to leave it as is, as it should be meaningful to most people.